sycl/doc/extensions/AsyncGroupCopy/async_group_copy.asciidoc

```diff
@@ -0,0 +1,115 @@
+# Async Group Copy Extension for DPC++: SYCL_EXT_ONEAPI_ASYNC_GROUP_COPY
+:source-highlighter: coderay
+:coderay-linenums-mode: table
+:dpcpp: pass:[DPC++]
+
+// This section needs to be after the document title.
+:doctype: book
+:toc2:
+:toc: left
+:encoding: utf-8
+:lang: en
+
+:blank: pass:[ +]
+
+// Set the default source code type in this document to C++,
+// for syntax highlighting purposes.  This is needed because
+// docbook uses c++ and html5 uses cpp.
+:language: {basebackend@docbook:c++:cpp}
+
+
+== Notice
+
+Copyright (c) 2021-2021 Intel Corporation.  All rights reserved.
+
+IMPORTANT: This specification is a draft.
+
+NOTE: Khronos(R) is a registered trademark and SYCL(TM) and SPIR(TM) are
+trademarks of The Khronos Group Inc.  OpenCL(TM) is a trademark of Apple Inc.
+used by permission by Khronos.
+
+This extension is written against the SYCL 2020 revision 3 specification.  All
+references below to the "core SYCL specification" or to section numbers in the
+SYCL specification refer to that revision.
+
+## Contributors
+
+* Gordon Brown
+* Tadej Ciglaric
+* Jack Kirk
+* Finlay Marno
+
+## Introduction
+
+This document proposes that `async_work_group_copy` be deprecated and replaced 
+with `async_group_copy`. `async_group_copy` is a non-member function, in line
+with the other group functions, which generalizes `async_work_group_copy` to
+also work with sub-groups.
+
+## Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification section 6.3.3 "Feature test macros". Therefore, an implementation
+supporting this extension must predefine the macro
+SYCL_EXT_ONEAPI_ASYNC_GROUP_COPY to one of the values defined in the table
+below. Applications can test for the existence of this macro to determine if the
+implementation supports this feature, or applications can test the macro’s value
+to determine which of the extension’s APIs the implementation supports.
+
+[%header,cols="1,5"]
+|===
+|Value |Description
+|1     |Initial extension implementation.
+|===
+
+
+## New `async_group_copy` function.
+Currently the `nd_item` and `group` classes have the member function
+`async_work_group_copy`. These will be deprecated in favor of a single function 
+`async_group_copy` that will have the group as a template parameter, much
+like `group_barrier`.
+
+`async_group_copy` methods would be valid for groups `group` and `sub_group`.
+Note that the destination and source arguments have been have been swapped to be
+more in line with other SYCL copy functions.
```

Comment:
I know this is inherited from SYCL 2020, but the fact that these operations only support copying from local to global memory (or vice versa) is quite subtle.  If there aren't any plans to generalize this, I think a short paragraph explaining that here would be helpful.

I'm also curious to hear your thoughts on whether generalizing this would be a good idea, or not.  What are the reasons not to support generic pointers here, or copies within an address space?

Reply:
Currently most of these arrangements would lead to just falling back on generic code since I don't think the backends have an optimized version. PTX only actually has an asynchronous copy from global to local and not the other way around.

I think this async method should only be used as a potential optimization, and not as the first choice for a basic implementation. I'd guess that in some cases this async method could even lead to overhead (though a benchmark would be required to see if that is true). Restraining it to somewhat model the real capabilities of hardware might help users not pick it first.

Reply:
That's a good argument, and I think adding some text to this effect in the proposal would help people understand the difference between this feature and a higher-level algorithm that might come later.

Part of me wonders whether there's anything we can do with the naming to make it clear that this is a low-level feature that may have hardware-specific considerations and map directly to certain hardware features.  We don't really have anything like that today -- the closest equivalent I can think of is the `native` namespace, which is currently only used for math operations.

Reply:
the optimization warning has been included. Sorry, I'm not sure about a better name.